// @expo/eas-build-job/dist/ios.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JobSchema = exports.imageMatchRules = exports.builderBaseImages = void 0;
const joi_1 = __importDefault(require("joi"));
const common_1 = require("./common");
const TargetCredentialsSchema = joi_1.default.object().keys({
    provisioningProfileBase64: joi_1.default.string().required(),
    distributionCertificate: joi_1.default.object({
        dataBase64: joi_1.default.string().required(),
        password: joi_1.default.string().allow('').required(),
    }).required(),
});
const BuildCredentialsSchema = joi_1.default.object().pattern(joi_1.default.string().required(), TargetCredentialsSchema);
exports.builderBaseImages = [
    'default',
    'latest',
    'stable',
    'macos-big-sur-11.4-xcode-12.5',
    'macos-big-sur-11.4-xcode-13.0',
    'macos-monterey-12.1-xcode-13.2',
    'macos-monterey-12.3-xcode-13.3',
    'macos-monterey-12.4-xcode-13.4',
    'macos-monterey-12.6-xcode-14.0',
    'macos-monterey-12.6-xcode-14.1',
];
exports.imageMatchRules = [
    {
        image: 'macos-big-sur-11.4-xcode-13.0',
        sdkSemverRange: '<=44',
        workflows: [common_1.Workflow.MANAGED],
    },
    {
        image: 'macos-monterey-12.4-xcode-13.4',
        sdkSemverRange: '>=45 <47',
        workflows: [common_1.Workflow.MANAGED],
    },
    {
        image: 'macos-monterey-12.6-xcode-14.0',
        sdkSemverRange: '>=47',
    },
    {
        image: 'macos-monterey-12.6-xcode-14.0',
        reactNativeSemverRange: '>=0.70.0',
    },
];
const BuilderEnvironmentSchema = joi_1.default.object({
    image: joi_1.default.string().valid(...exports.builderBaseImages),
    node: joi_1.default.string(),
    yarn: joi_1.default.string(),
    expoCli: joi_1.default.string(),
    bundler: joi_1.default.string(),
    fastlane: joi_1.default.string(),
    cocoapods: joi_1.default.string(),
    env: common_1.EnvSchema,
});
exports.JobSchema = joi_1.default.object({
    mode: joi_1.default.string()
        .valid(...Object.values(common_1.BuildMode))
        .default(common_1.BuildMode.BUILD),
    type: joi_1.default.when('mode', {
        is: joi_1.default.string().valid(common_1.BuildMode.BUILD),
        then: joi_1.default.string()
            .valid(...Object.values(common_1.Workflow))
            .required(),
        otherwise: joi_1.default.string().valid(common_1.Workflow.UNKNOWN).default(common_1.Workflow.UNKNOWN),
    }),
    triggeredBy: joi_1.default.string()
        .valid(...Object.values(common_1.BuildTrigger))
        .default(common_1.BuildTrigger.EAS_CLI),
    projectArchive: common_1.ArchiveSourceSchema.required(),
    resign: joi_1.default.when('mode', {
        is: joi_1.default.string().valid(common_1.BuildMode.BUILD),
        then: joi_1.default.any().strip(),
        otherwise: joi_1.default.object({
            applicationArchiveSource: common_1.ArchiveSourceSchema.required(),
        }).required(),
    }),
    platform: joi_1.default.string().valid(common_1.Platform.IOS).required(),
    projectRootDirectory: joi_1.default.when('mode', {
        is: joi_1.default.string().valid(common_1.BuildMode.BUILD),
        then: joi_1.default.string().required(),
        otherwise: joi_1.default.any().strip(),
    }),
    buildProfile: joi_1.default.when('mode', {
        is: joi_1.default.string().valid(common_1.BuildMode.BUILD),
        then: joi_1.default.when('triggeredBy', {
            is: joi_1.default.string().valid(common_1.BuildTrigger.GIT_BASED_INTEGRATION),
            then: joi_1.default.string().required(),
            otherwise: joi_1.default.string(),
        }),
        otherwise: joi_1.default.string(),
    }),
    releaseChannel: joi_1.default.string(),
    updates: joi_1.default.object({
        channel: joi_1.default.string(),
    }),
    secrets: joi_1.default.object({
        buildCredentials: BuildCredentialsSchema,
        environmentSecrets: common_1.EnvironmentSecretsSchema,
        robotAccessToken: joi_1.default.string(),
    }).required(),
    builderEnvironment: BuilderEnvironmentSchema,
    cache: common_1.CacheSchema.default(),
    developmentClient: joi_1.default.boolean(),
    simulator: joi_1.default.boolean(),
    version: joi_1.default.object({
        buildNumber: joi_1.default.string(),
    }),
    buildArtifactPaths: joi_1.default.array().items(joi_1.default.string()),
    scheme: joi_1.default.string(),
    buildConfiguration: joi_1.default.string(),
    applicationArchivePath: joi_1.default.string(),
    username: joi_1.default.string(),
    experimental: joi_1.default.object({
        prebuildCommand: joi_1.default.string(),
    }),
}).oxor('releaseChannel', 'updates.channel');
