// @expo/eas-json/build/build/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BuildProfileSchema = void 0;
const tslib_1 = require("tslib");
const joi_1 = tslib_1.__importDefault(require("joi"));
const semver_1 = tslib_1.__importDefault(require("semver"));
const types_1 = require("./types");
const AllowedCommonResourceClasses = [types_1.ResourceClass.DEFAULT, types_1.ResourceClass.MEDIUM];
const AllowedAndroidResourceClasses = [
    ...AllowedCommonResourceClasses,
    types_1.ResourceClass.LARGE,
];
const AllowedIosResourceClasses = [
    ...AllowedCommonResourceClasses,
    types_1.ResourceClass.M1_MEDIUM,
    types_1.ResourceClass.INTEL_MEDIUM,
];
const CacheSchema = joi_1.default.object({
    disabled: joi_1.default.boolean(),
    key: joi_1.default.string().max(128),
    cacheDefaultPaths: joi_1.default.boolean(),
    customPaths: joi_1.default.array().items(joi_1.default.string()),
});
const CommonBuildProfileSchema = joi_1.default.object({
    credentialsSource: joi_1.default.string().valid('local', 'remote').default('remote'),
    distribution: joi_1.default.string().valid('store', 'internal').default('store'),
    cache: CacheSchema,
    releaseChannel: joi_1.default.string().regex(/^[a-z\d][a-z\d._-]*$/),
    channel: joi_1.default.string().regex(/^[a-z\d][a-z\d._-]*$/),
    developmentClient: joi_1.default.boolean(),
    prebuildCommand: joi_1.default.string(),
    buildArtifactPaths: joi_1.default.array().items(joi_1.default.string()),
    node: joi_1.default.string().empty(null).custom(semverCheck),
    yarn: joi_1.default.string().empty(null).custom(semverCheck),
    expoCli: joi_1.default.string().empty(null).custom(semverCheck),
    env: joi_1.default.object().pattern(joi_1.default.string(), joi_1.default.string().empty(null)),
    autoIncrement: joi_1.default.alternatives().try(joi_1.default.boolean()),
    resourceClass: joi_1.default.string().valid(...AllowedCommonResourceClasses),
});
const AndroidBuildProfileSchema = CommonBuildProfileSchema.concat(joi_1.default.object({
    credentialsSource: joi_1.default.string().valid('local', 'remote'),
    distribution: joi_1.default.string().valid('store', 'internal'),
    withoutCredentials: joi_1.default.boolean(),
    image: joi_1.default.string(),
    ndk: joi_1.default.string().empty(null).custom(semverCheck),
    autoIncrement: joi_1.default.alternatives().try(joi_1.default.boolean(), joi_1.default.string().valid('version', 'versionCode')),
    resourceClass: joi_1.default.string().valid(...AllowedAndroidResourceClasses),
    artifactPath: joi_1.default.string(),
    applicationArchivePath: joi_1.default.string(),
    buildArtifactPaths: joi_1.default.array().items(joi_1.default.string()),
    gradleCommand: joi_1.default.string(),
    buildType: joi_1.default.string().valid('apk', 'app-bundle'),
}).oxor('artifactPath', 'applicationArchivePath'));
const IosBuildProfileSchema = CommonBuildProfileSchema.concat(joi_1.default.object({
    credentialsSource: joi_1.default.string().valid('local', 'remote'),
    distribution: joi_1.default.string().valid('store', 'internal'),
    enterpriseProvisioning: joi_1.default.string().valid('adhoc', 'universal'),
    autoIncrement: joi_1.default.alternatives().try(joi_1.default.boolean(), joi_1.default.string().valid('version', 'buildNumber')),
    simulator: joi_1.default.boolean(),
    resourceClass: joi_1.default.string().valid(...AllowedIosResourceClasses),
    image: joi_1.default.string(),
    bundler: joi_1.default.string().empty(null).custom(semverCheck),
    fastlane: joi_1.default.string().empty(null).custom(semverCheck),
    cocoapods: joi_1.default.string().empty(null).custom(semverCheck),
    artifactPath: joi_1.default.string(),
    applicationArchivePath: joi_1.default.string(),
    buildArtifactPaths: joi_1.default.array().items(joi_1.default.string()),
    scheme: joi_1.default.string(),
    buildConfiguration: joi_1.default.string(),
}).oxor('artifactPath', 'applicationArchivePath'));
exports.BuildProfileSchema = CommonBuildProfileSchema.concat(joi_1.default.object({
    extends: joi_1.default.string(),
    android: AndroidBuildProfileSchema,
    ios: IosBuildProfileSchema,
}));
function semverCheck(value) {
    if (semver_1.default.valid(value)) {
        return value;
    }
    else {
        throw new Error(`${value} is not a valid version`);
    }
}
