// @expo/eas-json/build/build/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceClass = exports.CredentialsSource = void 0;
const eas_build_job_1 = require("@expo/eas-build-job");
var CredentialsSource;
(function (CredentialsSource) {
    CredentialsSource["LOCAL"] = "local";
    CredentialsSource["REMOTE"] = "remote";
})(CredentialsSource = exports.CredentialsSource || (exports.CredentialsSource = {}));
var ResourceClass;
(function (ResourceClass) {
    ResourceClass["DEFAULT"] = "default";
    ResourceClass["LARGE"] = "large";
    /**
     * @deprecated use M1_MEDIUM instead
     * @experimental
     * This resource class is not yet ready to be used in production. For testing purposes only. Might be deprecated / deleted at any time.
     */
    ResourceClass["M1_EXPERIMENTAL"] = "m1-experimental";
    ResourceClass["M1_MEDIUM"] = "m1-medium";
    ResourceClass["M1_LARGE"] = "m1-large";
    ResourceClass["INTEL_MEDIUM"] = "intel-medium";
    ResourceClass["MEDIUM"] = "medium";
})(ResourceClass = exports.ResourceClass || (exports.ResourceClass = {}));
