// @expo/eas-json/build/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EasJsonUtils = void 0;
const resolver_1 = require("./build/resolver");
const errors_1 = require("./errors");
const resolver_2 = require("./submit/resolver");
class EasJsonUtils {
    static async getBuildProfileNamesAsync(accessor) {
        var _a;
        const easJson = await accessor.readAsync();
        return Object.keys((_a = easJson === null || easJson === void 0 ? void 0 : easJson.build) !== null && _a !== void 0 ? _a : {});
    }
    static async getBuildProfileAsync(accessor, platform, profileName) {
        const easJson = await accessor.readAsync();
        return (0, resolver_1.resolveBuildProfile)({ easJson, platform, profileName });
    }
    static async getCliConfigAsync(accessor) {
        var _a;
        try {
            const easJson = await accessor.readAsync();
            return (_a = easJson.cli) !== null && _a !== void 0 ? _a : null;
        }
        catch (err) {
            if (err instanceof errors_1.MissingEasJsonError) {
                return null;
            }
            throw err;
        }
    }
    static async getSubmitProfileNamesAsync(accessor) {
        var _a;
        const easJson = await accessor.readAsync();
        return Object.keys((_a = easJson === null || easJson === void 0 ? void 0 : easJson.submit) !== null && _a !== void 0 ? _a : {});
    }
    static async getSubmitProfileAsync(accessor, platform, profileName) {
        const easJson = await accessor.readAsync();
        return (0, resolver_2.resolveSubmitProfile)({ easJson, platform, profileName });
    }
}
exports.EasJsonUtils = EasJsonUtils;
